File: Juniper/34.在排序数组中查找元素的第一个和最后一个位置.py
#
# @lc app=leetcode.cn id=34 lang=python3
#
# [34] 在排序数组中查找元素的第一个和最后一个位置
#
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        if not nums :
            return [-1,-1]
        lens = len(nums)
        if lens == 1 and nums[0] == target:
            return [0,0]
        low, high = 0, lens-1
        flag = 0
        while low <= high:
            mid =  (low + high)//2
            if nums[mid] == target:
                flag = 1
                break
            elif nums[mid] < target:
                low =mid +1
            else :
                high =mid-1
        if flag:
            left , right = mid, mid
            while left >= 0 and nums[left] == target:
                left -=1
            while right <=(lens-1) and nums[right] == target :
                right +=1
        else :
            return [-1,-1] 
        return [left+1,right-1]

File: hahlw/array/34.在排序数组中查找元素的第一个和最后一个位置.cpp
/*
 * @lc app=leetcode.cn id=34 lang=cpp
 *
 * [34] 在排序数组中查找元素的第一个和最后一个位置
 */
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        vector<int >res{-1,-1};
        //往前搜索：
        int lo=0,hi=nums.size()-1;
        int t=-1,e=-1;
        while(lo<=hi){
            int mid=lo+(hi-lo)/2;
            if(nums[mid]==target){
                t=mid;
                hi=mid-1;
            }
            else if(nums[mid]>target)hi=mid-1;
            else lo=mid+1;
        }
        if(t==-1)return res;
        //向后
        lo=0,hi=nums.size()-1;
        while(lo<=hi){
            int mid=lo+(hi-lo)/2;
            if(nums[mid]==target){
                e=mid;
                lo=mid+1;
            }
            else if(nums[mid]>target)hi=mid-1;
            else lo=mid+1;
        }
        res[0]=t;
        res[1]=e;
        return res;
    }
};

